### PR TITLE
Branch lesson11 12

### PR DESCRIPTION
**Pull request by TatsianaKazachenka**

### CODE REVIEW

File: src/main/java/pages/CartPage.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+import java.util.List;
+
+public class CartPage extends BasePage {
+    public static final String CART_ITEM = "//*[text() = '%s']/ancestor::*[@class='cart_item']";
```
Comment (TatsianaKazachenka) [inline src/main/java/pages/CartPage.java:12]: how can I change to @FindBy here if there is '% s'?

File: src/main/java/pages/LoginPage.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    public LoginPage openPage() {
+        load();
+        isLoaded();
+        return this;
+    }
+
+    public List<String> getListLogin(){
+        return Arrays.asList(listLogin.getText()
+                .replace("Accepted usernames are:\n", "").split("\n"));
```
Comment (TatsianaKazachenka) [inline src/main/java/pages/LoginPage.java:63]: how to get all possible logins into an array and avoid "Accepted usernames are: \ n" and .split ("\ n")?

File: src/main/java/pages/CartPage.java (hunk 1 above)
Comment (vchesnakoutest) [reply src/main/java/pages/CartPage.java:12]: no way :) just as you did, define a String

File: src/main/java/pages/CartPage.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    public CartPage(WebDriver driver) {
+        super(driver);
+    }
+
+    @Override
+    public CartPage waitForPageOpened() {
+        isLoaded();
+        return this;
+    }
+    /**
```
Comment (vchesnakoutest) [inline src/main/java/pages/CartPage.java:38]: add one empty line before method declaration

File: src/main/java/pages/CartPage.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    public boolean isEmptyCart() {
+        try {
+            return cartEmptyItems.isDisplayed();
+        } catch (Exception e) {
+            return false;
+        }
+    }
+    public void clickContinueShopping(){
```
Comment (vchesnakoutest) [inline src/main/java/pages/CartPage.java]: add one empty line before method declaration

File: src/main/java/pages/LoginPage.java (hunk 2 above)
Comment (vchesnakoutest) [reply src/main/java/pages/LoginPage.java:63]: in this case it's not easy to do that, you did it correctly using .replace. That is acceptable approach.

File: src/main/java/pages/LoginPage.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+    public boolean isShowErrorMessage() {
+        try {
+            return errorMessage.isDisplayed();
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    public boolean isEmptyField() {
```
Comment (vchesnakoutest) [inline src/main/java/pages/LoginPage.java]: isUsernamePasswordFieldsEmpty() - be more specific

File: src/main/java/pages/LoginPageFactory.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        return null;
+    }
+
+    @FindBy(xpath="//*[@data-test='username']")
+    WebElement usernameInput;
+    @FindBy(xpath="//*[@data-test='password']")
+    WebElement userpasswordInput;
+
+    @FindBy(id = "login-button")
+    WebElement loginBtn;
```
Comment (vchesnakoutest) [inline src/main/java/pages/LoginPageFactory.java]: better to move findBy's above all methods in class. and delete extra space between userpasswordInput and loginBtn

File: src/main/java/pages/LoginPageFactory.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+     * Open login page
+     */
+    public void openPage() {
+        String loginUrl = BASE_URL  + LOGIN_URL;
+        driver.get(loginUrl);
+    }
+
+    @Override
+    protected void load() {
+
```
Comment (vchesnakoutest) [inline src/main/java/pages/LoginPageFactory.java]: try not to keep empty methods, you can add comment like:
//TODO: Implement this method

File: src/main/java/pages/MenuPage.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+                element = hrefAllItems;
+                break;
+            case "about":
+                element = hrefAbout;
+                break;
+            case "logout":
+                element = hrefLogout;
+                break;
+            default:
+                break;
```
Comment (vchesnakoutest) [inline src/main/java/pages/MenuPage.java]: for default value you can throw an Exception with message like "There is not such type '%s' !", type

In other way if type is unknown you will see NullPointerException which is not informative in this case

File: src/main/java/pages/ProductsPage.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    public List<WebElement> getProductElements() {
+        return productsItem;
+    }
+
+    /**
+     * sorting the product list
+     *
+     * @param typeSort
+     */
+    public void clickSortProducts(String typeSort) {
```
Comment (vchesnakoutest) [inline src/main/java/pages/ProductsPage.java:94]: make methods return object of ProductsPage if we use 'Chain of invocation' pattern

File: src/test/java/tests/CartTests.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class CartTests extends BaseTests {
+    @Test
+    public void removeItemCartTest() {
+        productsPage.openPage().waitForPageOpened();
+        productsPage.clickAddOrRemoveProductItem(PRODUCT_NAME, true);
+        cartPage.openPage().waitForPageOpened();
+        cartPage.clickRemoveBtn();
```
Comment (vchesnakoutest) [inline src/test/java/tests/CartTests.java]: make test in 'Chain of invocations' style:

```
        productsPage.openPage()
                .waitForPageOpened()
                .clickAddOrRemoveProductItem(PRODUCT_NAME, true);
        cartPage.openPage().
                waitForPageOpened().
                clickRemoveBtn();
```

File: src/test/java/tests/CartTests.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        Assert.assertFalse(cartPage.isEmptyCart());
+    }
+
+    @Test
+    public void continueShoppingTest() {
+        cartPage.openPage().waitForPageOpened();
+        cartPage.clickContinueShopping();
+        Assert.assertTrue(productsPage.isPageOpened());
```
Comment (vchesnakoutest) [inline src/test/java/tests/CartTests.java:23]: and others as well

File: src/main/java/pages/CartPage.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+     *
+     * @param productName
+     * @return
+     */
+    public Boolean isCheckedProduct(String productName) {
+        List<WebElement> elements = cartItems;
+        boolean isPoduct = false;
+        for (WebElement element : elements) {
+            isPoduct = element.getText().equals(productName);
+            if (isPoduct) continue;
```
Comment (vchesnakoutest) [inline src/main/java/pages/CartPage.java]: perhaps this 'if' is not necessary

File: src/test/java/tests/MenuTests.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        menuPage.clickHref("logout");
+        Assert.assertTrue(loginPage.isEmptyField());
+    }
+
+    @Test
+    public void closeMenuTest(){
+        menuPage.openPage().waitForPageOpened();
+        menuPage.clickCloseBtn().isWaitForAboutPageOpened();
+        Assert.assertFalse(menuPage.isShowMenu());
+    }
```
Comment (vchesnakoutest) [inline src/test/java/tests/MenuTests.java]: this test fails with ElementClickInterceptedException

File: src/test/java/tests/ProductsTests.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.*;
+
+public class ProductsTests extends BaseTests {
+    public static final String PRODUCT_TYPE_SORT = "az";
+
+    @Test
+    public void addProductToCartTest() {
+        login();
+        productsPage.clickAddOrRemoveProductItem(PRODUCT_NAME, true);
+
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTests.java:16]: remove empty lines

File: src/test/java/tests/ProductsTests.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+        ArrayList<String> elementsText = new ArrayList<String>();
+        for (WebElement element : elements) {
+            elementsText.add(element.getText());
+        }
+        ArrayList<String> sortElementsText = elementsText;
+        Collections.sort(sortElementsText, new Comparator<String>() {
+            public int compare(String s1, String s2) {
+                return s1.toString().compareTo(s2.toString());
+            }
+        });
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTests.java]: better to move this logic out from Test and place in Page class. Just return necessary collections

**Review by vchesnakoutest — COMMENTED** [review]
please see comments and make fixes

File: src/main/java/pages/CartPage.java (hunk 3 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/CartPage.java:38]: done

File: src/main/java/pages/CartPage.java (hunk 4 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/CartPage.java]: done

File: src/main/java/pages/LoginPageFactory.java (hunk 6 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/LoginPageFactory.java]: done

File: src/test/java/tests/ProductsTests.java (hunk 14 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTests.java:16]: done

File: src/main/java/pages/LoginPageFactory.java (hunk 7 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/LoginPageFactory.java]: I apologize. there was no need to add this file, all the logic from this file was moved to the LoginPage file

File: src/main/java/pages/MenuPage.java (hunk 8 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/MenuPage.java]: done

File: src/main/java/pages/ProductsPage.java (hunk 9 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/ProductsPage.java:94]: done

File: src/test/java/tests/CartTests.java (hunk 10 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/CartTests.java]: done

File: src/test/java/tests/CartTests.java (hunk 11 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/CartTests.java:23]: done

File: src/main/java/pages/LoginPage.java (hunk 5 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/LoginPage.java]: done

File: src/test/java/tests/MenuTests.java (hunk 13 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/MenuTests.java]: done

File: src/test/java/tests/ProductsTests.java (hunk 15 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTests.java]: done

File: src/main/java/pages/CartPage.java (hunk 12 above)
Comment (TatsianaKazachenka) [reply src/main/java/pages/CartPage.java]: done

**Review by vchesnakoutest — APPROVED** [review]

**Pull request merged**